api.go

```diff
@@ -0,0 +1,469 @@
+	//   "stop_sequence": The response reached a stop sequence.
+	//   "tools_use": The model invoked one or more tools.
+	StopReason string `json:"stop_reason"`
+	// StopSequence is the stop sequence that caused the response to stop,
+	// in the case of a StopReason of "stop_sequence"
+	StopSequence *string `json:"stop_sequence"`
+	// Usage is the usage statistics for the response.
+	Usage struct {
+		InputTokens  int `json:"input_tokens"`
+		OutputTokens int `json:"output_tokens"`
+	}
+}
+
+// SampleSettings is used to set the settings for the request. Usually it
+// pertains to samplers.
+type SampleSettings struct {
+	// Model to use for the sample.
+	Model string `json:"model"`
+	// Version of the model to use for the sample.
+	Version string `json:"version"`
+	// Beta is the beta version of the model to use for the sample.
+	Beta string `json:"beta"`
+	// MaxTokens is the mas number of tokens to generate.
+	MaxTokens int `json:"max_tokens"`
+	// The temperature to use for sampling.
+	Temperature float64 `json:"temperature"`
+}
+
+// A Conversation is a sequence of messages between a user and an assistant.
+type Conversation struct {
+	// System is the system prompt to use for the conversation.
+	System *string
+	// Messages is the sequence of messages in the conversation. This always
+	// starts with a user message, and alternates between user and assistant.
+	Messages *[]*Message
+	// Usage is the usage statistics for the conversation in total.
+	Usage struct {
+		InputTokens  int `json:"input_tokens"`
+		OutputTokens int `json:"output_tokens"`
+	}
+	// HttpClient is the HTTP client used for API requests
+	HttpClient *http.Client
+	// apiToken is the API token used for API requests
+	ApiToken string `json:"-"`
+	// Settings is the settings for the conversation.
+	Settings *SampleSettings
+}
+
+// NewConversation creates a new conversation with the given system prompt. It
+// initializes the messages and usage statistics, as well as reasonable
+// defaults.
+func NewConversation(system string) *Conversation {
+	messages := make([]*Message, 0)
+	conversation := Conversation{
+		System:   &system,
+		Messages: &messages,
+	}
+	conversation.Usage.OutputTokens = 0
+	conversation.Usage.InputTokens = 0
+	// Copy our default settings
+	settings := DefaultSettings
+	conversation.Settings = &settings
+	conversation.ApiToken = DefaultApiToken
+
+	// Initialize the HTTP client
+	conversation.HttpClient = &http.Client{}
+	return &conversation
+}
+
+// AddMessage adds a message to the conversation with the given role and
+// content. It used internally, and also can be used externally to
+// manipulate conversations.
+func (conversation *Conversation) AddMessage(
+	role string,
+	content *[]ContentBlock,
+) {
+	message := Message{
+		Role:    role,
+		Content: content,
+	}
+	*conversation.Messages = append(*conversation.Messages, &message)
+}
+
+// Messages API URI
+// POST /v1/messages
+var messagesURI = "https://api.anthropic.com/v1/messages"
```

Comment:
API base url being configurable would be useful for backends that act as proxies, or local inference runners that expose a claude-compatible api.

Reply:
Yeah, I could either:
* Change that to `MessagesApiURI` -- and expose it so that it's settable as `anthropic.MessagesApiURI`.
* Or I could have that on the `Conversation` object as well?
Thoughts?

Reply:
The pattern is already set for the former (via `DefaultApiToken` and `DefaultSettings`), so I'd follow with... both? (or that could be added to Settings)

I'd suggest it be left as the base url rather than messages, for extensibility's sake i.e configure as `"https://api.anthropic.com"`